### PR TITLE
[ACA-3358] Show user friendly error templates when there are no applications/processDefinitions

### PR DESCRIPTION
**Pull request by sivakumar414ram**
**Please check if the PR fulfills these requirements**

> - [x] The commit message follows our [guidelines](https://github.com/Alfresco/alfresco-ng2-components/wiki/Commit-format)
> - [x] Tests for the changes have been added (for bug fixes / features)
> - [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

> - [ ] Bugfix
> - [x] Feature
> - [ ] Code style update (formatting, local variables)
> - [ ] Refactoring (no functional changes, no api changes)
> - [ ] Build related changes
> - [ ] Documentation
> - [ ] Other... Please describe:

**What is the current behaviour?** (You can also link to an open issue here)

https://issues.alfresco.com/jira/browse/ACA-3358

**What is the new behaviour?**

**Does this PR introduce a breaking change?** (check one with "x")

> - [ ] Yes
> - [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: lib/process-services/src/lib/process-list/components/start-process.component.spec.ts [hunk 1]
```diff
@@ -747,4 +709,104 @@ describe('StartFormComponent', () => {
...
+            getDefinitionsSpy.and.returnValue(of([]));
+            getDeployedApplicationsSpy = spyOn(appsProcessService, 'getDeployedApplications').and.returnValue(of([]));
+        });
+
+        it('should show no process definition available template when application/process definitions are empty', () => {
+            component.showSelectApplicationDropdown = true;
+            const change = new SimpleChange(null, 3, true);
+            component.ngOnChanges({ 'appId': change });
+            component.applications = [];
+            component.processDefinitions = [];
```
Comment (arditdomi) [inline lib/process-services/src/lib/process-list/components/start-process.component.spec.ts:726]: You are spying on the getDeployedApplications and getDefinitions and returning empty array so the `component.applications` and the `component.processDefinitions` should already be empty right?

File: lib/process-services/src/lib/process-list/components/start-process.component.spec.ts [hunk 2]
```diff
@@ -747,4 +709,104 @@ describe('StartFormComponent', () => {
...
+            fixture.detectChanges();
+            const noProcessElement = fixture.nativeElement.querySelector('.adf-empty-content__title');
+            expect(noProcessElement).not.toBeNull('Expected no available process message to be present');
+            expect(noProcessElement.innerText.trim()).toBe('ADF_PROCESS_LIST.START_PROCESS.NO_PROCESS_DEFINITIONS');
+        });
+
+        it('should show no process definition available template if processDefinitions are empty', () => {
+            const change = new SimpleChange(null, 3, true);
+            component.ngOnChanges({ 'appId': change });
+            component.processDefinitions = [];
```
Comment (arditdomi) [inline lib/process-services/src/lib/process-list/components/start-process.component.spec.ts:736]: same here

File: lib/process-services/src/lib/process-list/components/start-process.component.spec.ts [hunk 3]
```diff
@@ -747,4 +709,104 @@ describe('StartFormComponent', () => {
...
+            const noProcessElement = fixture.nativeElement.querySelector('.adf-empty-content__title');
+            expect(noProcessElement).not.toBeNull('Expected no available process message to be present');
+            expect(noProcessElement.innerText.trim()).toBe('ADF_PROCESS_LIST.START_PROCESS.NO_PROCESS_DEF_SELECTED');
+        });
+
+        it('should show no start form template if selected process definition does not have start form', () => {
+            component.showSelectApplicationDropdown = true;
+            const change = new SimpleChange(null, 3, true);
+            component.ngOnChanges({ 'appId': change });
+            component.selectedProcessDef = testMultipleProcessDefs[0];
```
Comment (arditdomi) [inline lib/process-services/src/lib/process-list/components/start-process.component.spec.ts:757]: same here. You could re-spy and return the testMultipleProcessDefs[0]

File: lib/process-services/src/lib/process-list/components/start-process.component.spec.ts [hunk 4]
```diff
@@ -747,4 +709,104 @@ describe('StartFormComponent', () => {
...
+
+        it('should emit error event in case loading process definitions failed', async(() => {
+            getDefinitionsSpy.and.returnValue(throwError(processDefError));
+            const errorSpy = spyOn(component.error, 'emit');
+            const change = new SimpleChange(null, 3, true);
+            component.ngOnChanges({ 'appId': change });
+            fixture.detectChanges();
+            fixture.whenStable().then(() => {
+                expect(errorSpy).toHaveBeenCalledWith(processDefError);
+            });
```
Comment (arditdomi) [inline lib/process-services/src/lib/process-list/components/start-process.component.spec.ts:783]: can we use await fixture.whenStable(); And then the expectation in the next line?

File: lib/process-services/src/lib/process-list/components/start-process.component.spec.ts [hunk 5]
```diff
@@ -747,4 +709,104 @@ describe('StartFormComponent', () => {
...
+        it('should emit error event in case loading applications failed', async(() => {
+            getDeployedApplicationsSpy = spyOn(appsProcessService, 'getDeployedApplications').and.returnValue(throwError(applicationsError));
+            const errorSpy = spyOn(component.error, 'emit');
+            component.showSelectApplicationDropdown = true;
+            const change = new SimpleChange(null, 3, true);
+            component.ngOnChanges({ 'appId': change });
+            fixture.detectChanges();
+            fixture.whenStable().then(() => {
+                expect(errorSpy).toHaveBeenCalledWith(applicationsError);
+            });
```
Comment (arditdomi) [inline lib/process-services/src/lib/process-list/components/start-process.component.spec.ts:795]: same here

File: lib/process-services/src/lib/process-list/components/start-process.component.spec.ts [hunk 6]
```diff
@@ -747,4 +709,104 @@ describe('StartFormComponent', () => {
...
+            const errorSpy = spyOn(component.error, 'emit');
+            const change = new SimpleChange(null, 3, true);
+            component.ngOnChanges({ 'appId': change });
+            component.selectedProcessDef = testProcessDef;
+            component.name = 'mock name';
+            component.startProcess();
+            fixture.detectChanges();
+            fixture.whenStable().then(() => {
+                expect(errorSpy).toHaveBeenCalledWith(startProcessError);
+            });
```
Comment (arditdomi) [inline lib/process-services/src/lib/process-list/components/start-process.component.spec.ts:809]: same here

File: lib/process-services/src/lib/process-list/components/start-process.component.ts [hunk 7]
```diff
@@ -315,13 +328,28 @@ export class StartProcessInstanceComponent implements OnChanges, OnInit, OnDestr
...
+        return !!(this.selectedProcessDef && this.selectedProcessDef.id);
     }
 
     isProcessDefinitionsEmpty(): boolean {
         return this.processDefinitions.length === 0;
     }
 
+    disableDropdownButton(): boolean {
+        return this.showSelectApplicationDropdown && !this.isAppSelected();
+    }
```
Comment (arditdomi) [inline lib/process-services/src/lib/process-list/components/start-process.component.ts:351]: dropdownButton sounds confusing. Is it a dropdown or a button?

File: docs/process-services/components/start-process.component.md [hunk 8]
```diff
@@ -195,6 +195,25 @@ You can use the `showSelectApplicationDropdown` property to Hide or show applica
...
+### Error handling
+
+When an error ocurred during start process, start process component will emit an error event, So that you can catch an error event in your component and act depends on error thrown. For an example like below.
+
+```html
+ <adf-start-process 
+      [appId]="YOUR_APP_ID"
+      [title]="'ADF_PROCESS_LIST.START_PROCESS.FORM.TITLE'"
+      [name]="PROCESS_NAME"
+      [error]="onError($event)">
```
[the remark this answers is not in the document]
Comment (mauriziovitale) [reply docs/process-services/components/start-process.component.md:207]: nice spot @sivakumar414ram ?

File: lib/process-services/src/lib/process-list/components/start-process.component.html [hunk 9]
```diff
@@ -1,116 +1,145 @@
...
-                            (click)="displayDropdown($event)">
-                            <mat-icon>arrow_drop_down</mat-icon>
-                        </button>
+
+<ng-container *ngIf="isLoading(); then showLoadingTemplate; else showStartProcessTemplate"></ng-container>
+    <ng-template #showLoadingTemplate>
+        <mat-spinner class="adf-start-process-loading"></mat-spinner>
+    </ng-template>
+<ng-template #showStartProcessTemplate>
+    <ng-container *ngIf="hasApplications() || hasProcessDefinitions() ; else showEmptyTemplate">
```
[the remark this answers is not in the document]
Comment (sivakumar414ram) [reply lib/process-services/src/lib/process-list/components/start-process.component.html:7]: There are two scenarios that components needs to handle.

1. Start Process without application dropdown
      This is a default behaviour,  in this case will have only process definitions. If no process definitions show empty template.

2. Start process with application dropdown.
     Applications will be enabled based on flag showSelectApplicationDropdown, in this case if applications are empty component should show empty template.

[the remark this answers is not in the document]
Comment (arditdomi) [reply lib/process-services/src/lib/process-list/components/start-process.component.html:7]: makes sense. Thank you.

File: lib/core/templates/empty-content/empty-content.component.scss [hunk 10]
```diff
@@ -35,4 +37,8 @@
             text-align: center;
         }
     }
+
+    .adf-icon {
+        opacity: 0.6;
```
[the remark this answers is not in the document]
Comment (sivakumar414ram) [reply lib/core/templates/empty-content/empty-content.component.scss]: Updated

File: docs/process-services/components/start-process.component.md [hunk 11]
```diff
@@ -195,6 +195,25 @@ You can use the `showSelectApplicationDropdown` property to Hide or show applica
 ![Start process with selected application](../../docassets/images/start-process-with-selected-application.png)
 
 
+### Error handling
+
+When an error ocurred during start process, start process component will emit an error event, So that you can catch an error event in your component and act depends on error thrown. For an example like below.
```
[the remark this answers is not in the document]
Comment (sivakumar414ram) [reply docs/process-services/components/start-process.component.md]: Updated

File: docs/process-services/components/start-process.component.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (sivakumar414ram) [reply docs/process-services/components/start-process.component.md]: Updated, Thanks @arditdomi

File: lib/process-services/src/lib/process-list/components/start-process.component.spec.ts (hunk 1 above)
Comment (sivakumar414ram) [reply lib/process-services/src/lib/process-list/components/start-process.component.spec.ts]: Updated, Thanks @arditdomi

File: demo-shell/src/app/components/process-service/process-service.component.ts [hunk 12]
```diff
@@ -403,6 +404,10 @@ export class ProcessServiceComponent implements AfterViewInit, OnDestroy, OnInit
         this.reloadProcessFilters();
     }
 
+    onStartProcessError(event: any) {
+        this.notificationService.openSnackMessage(event.response.body.message);
```
Comment (arditdomi) [inline demo-shell/src/app/components/process-service/process-service.component.ts:408]: can we use the https://github.com/Alfresco/alfresco-ng2-components/blob/develop/lib/core/notifications/services/notification.service.ts#L74 ?

File: lib/process-services/src/lib/process-list/components/start-process.component.spec.ts [hunk 13]
```diff
@@ -747,4 +709,107 @@ describe('StartFormComponent', () => {
             expect(processNameInput.disabled).toEqual(false);
         });
    });
+
+    describe('Empty Template', () => {
+
+        it('should show no process definition available template when application/process definitions are empty', async() => {
+            getDeployedApplicationsSpy = spyOn(appsProcessService, 'getDeployedApplications').and.returnValue(of([]));
+            getDefinitionsSpy.and.returnValue(of([]));
```
Comment (eromano) [inline lib/process-services/src/lib/process-list/components/start-process.component.spec.ts:717]: https://w3.alfresco.com/confluence/pages/viewpage.action?spaceKey=AP&title=Code+Style+E2E  Style 05-14
 to dense

File: lib/process-services/src/lib/process-list/components/start-process.component.ts [hunk 14]
```diff
@@ -93,11 +93,11 @@ export class StartProcessInstanceComponent implements OnChanges, OnInit, OnDestr
 
     /** Emitted when the process is canceled. */
     @Output()
-    cancel: EventEmitter<ProcessInstance> = new EventEmitter<ProcessInstance>();
+    cancel: EventEmitter<any> = new EventEmitter<any>();
```
Comment (eromano) [inline lib/process-services/src/lib/process-list/components/start-process.component.ts:96]: any seems to be too generic

Comment (sivakumar414ram) [reply lib/process-services/src/lib/process-list/components/start-process.component.ts]: Updated

File: lib/process-services/src/lib/process-list/components/start-process.component.spec.ts (hunk 13 above)
Comment (sivakumar414ram) [reply lib/process-services/src/lib/process-list/components/start-process.component.spec.ts:717]: I will go through the above confluence page, and next time will follow guide lines. Thanks :+1:

**Pull request merged**